### PR TITLE
Make binary classifier work with BCELoss

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Until now, `NeuralNetBinaryClassifier` assumes that `BCEWithLogitLoss` is used. Now, `BCELoss` works as well.

Note: Unless there is a very good reason not to, it's still recommended to use `NeuralNetClassifier`, even when dealing with a binary classification problem.

### CODE REVIEW

File: skorch/utils.py [hunk 1]
```diff
@@ -632,7 +633,7 @@ def _infer_predict_nonlinearity(net):
     if isinstance(criterion, CrossEntropyLoss):
         return partial(torch.softmax, dim=-1)
 
-    if isinstance(criterion, BCEWithLogitsLoss):
+    if isinstance(criterion, (BCEWithLogitsLoss, BCELoss)):
```
Comment (thomasjpfan) [inline skorch/utils.py:636]: In the test, `module__output_nonlin=torch.nn.Sigmoid()`, which means the output already has sigmoid applied. I think using this non-linearity would apply another sigmoid.

If we assume that the module already does sigmoid then the module already outputs probabilities. In that case, I think the non-linearity should only be `torch.stack((1 - prob, prob), 1)` part of `_sigmoid_then_2d`.

Comment (BenjaminBossan) [reply skorch/utils.py]: Very good catch, thanks. I fixed the bug by not applying the sigmoid in the postprocessing step in `utils.py`. We still need it as output nonlinearity because that is what `BCELoss` expects.

To catch this kind of bug in the future, I extended the unit tests. When sigmoid is applied twice, all probabilities would be in the range of [sigmoid(-1), sigmoid(1)]. I now check that at least some returned probabilities are outside that range.

File: skorch/tests/test_classifier.py [hunk 2]
```diff
@@ -255,6 +255,11 @@ def test_predict_predict_proba(self, net, data, threshold):
         y_pred_proba = net.predict_proba(X)
         assert y_pred_proba.shape == (X.shape[0], 2)
 
+        # the tests below check that we don't accidentally apply sigmoid twice,
+        # which would result in all probas being withiin [0.2689, 0.7311].
+        assert (y_pred_proba < 0.26).any()
```
Comment (thomasjpfan) [inline skorch/tests/test_classifier.py:260]: Is this suppose to be `expit(0) = 0.5` on the lower bound? ([scipy.special.expit](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.expit.html))

```suggestion
        assert (y_pred_proba < expit(0)).any()
```

Comment (BenjaminBossan) [reply skorch/tests/test_classifier.py]: Indeed, we are comparing to `expit(1)` and `expit(-1)`. The loewr bound is not `expit(0)` because we create the probabilities as `torch.stack((1 - prob, prob), 1)`, so the lowest value that could be achieved by applying sigmoid twice is `1 - expit(1)`, which is equal to `expit(-1)`.

Comment (thomasjpfan) [reply skorch/tests/test_classifier.py]: Can you add this information in a code comment?

**Comment by BenjaminBossan** [comment]
Note: I suspect CI is failing because of the new `accelerate` release, I'll investigate.

**Review by thomasjpfan — APPROVED** [review]
LGTM

The [failing tests](https://github.com/skorch-dev/skorch/runs/7449764358?check_suite_focus=true) do not look related to this PR, but they look concerning.
